designdocs/user-handles.md

```diff
@@ -0,0 +1,316 @@
+* Store ```msa_id``` to ```handle``` mapping and ```handle``` to ```msa_id``` mapping.
+
+### Handling Race Conditions
+
+There is an exceedingly small chance that two users may attempt to claim the same handle and suffix combination within the same block. In such cases, the transaction will fail to create the MSA, and the app will receive a failure for the transaction rather than the expected MsaHandleCreated event.
+
+## Sequence Diagram
+
+```mermaid
+sequenceDiagram
+    participant Wallet
+    actor User
+    participant App
+    participant RPC
+    participant Frequency
+    User->>App: enter desired handle
+    App->>Wallet: send signup data
+    Wallet->>Wallet: perform 2FA if first attempt
+    Wallet-->>App: return signed payload
+    App->>Frequency: create_msa_handle(.., signed_payload)
+    Frequency-->>Frequency: query current seed
+    Frequency-->>Frequency: compute suffix options
+    Frequency-->>Frequency: check_handle_availability(...handle, suffix)
+    Frequency-->>Frequency: submit transaction
+    Frequency-->>App: MsaHandleCreated event with (msa_id, handle+suffix) 
+    App->>User: proceed with setup
+    User->>App: request to get msa_id
+    App->>Frequency: query state for msa_id (handle<->msa_id)
+    Frequency-->>App: return msa_id (State Query handle<->msa_id)
+    User->>App: request to get handle for msa_id
+    App->>RPC: get_handle_for_msa_id(msa_id)
+    RPC->>Frequency: query for handle given msa_id (index handle<->msa_id)
+    Frequency-->>RPC: return handle for msa_id
+    RPC-->>App: return handle for msa_id
+```
+
+## Storage
+
+* **HandleSuffixMin**: This storage value (```u32``) will keep track of the minimum suffix value allowed for a given handle.
+* **HandleSuffixMax**: This storage value (```u32``) will keep track of the maximum suffix value allowed for a given handle.
+* **HandleRetirementPeriod**: This storage value (```u32``) will keep track of the number of blocks a handle must be retired before it can be reused.
+* **MSAIdToUserHandle**: This storage will map MSAIds to their corresponding user handles.
+* **UserHandlesToMSAId**: This storage will map user handles to their corresponding MSAIds.
+* **RetiredHandles**: This storage will keep track of all retired handles and the block number at which they were retired. This probably is a spike to run few ideas on how to implement this.
+* **UsedSuffixes**: This storage will keep track of all used suffixes to ensure that no two handles have the same suffix. (May not be necessary)
+* **Seed (current and previous)**: This storage values (```tuple (s0, s1)```) will keep track of the current and previous seed values.
+
+**Note:** The handle being mapped to MSA ID is the handle with the suffix i.e. full handle.
+
+## Required Extrinsics
+
+### Primitives
+
+``` rust
+
+MsaHandlePayload {
+    handle: &[u8],
+    suffix: Option<u32>,
+}
+```
+
+### Create user handle with chosen handle and optional suffix
```

Comment:
We should mention somewhere that `create`, and `change` (and `retire` depending on if it is Pays::No) should be Capacity possible transactions.

Reply:
Making these Capacity transactions might incentive providers to create MSA ids and handles. These could then be resold and create a market for them. However, given that we do not have any punishment system, we put too much trust in providers not to exploit.

Reply:
Added notes and assumption for the same , thanks